core/src/main/java/com/linecorp/armeria/server/RoutingTrie.java

```diff
@@ -97,16 +115,80 @@ private RoutingTrie(Node<V> root) {
     @VisibleForTesting
     Node<V> findNode(String path, boolean exact) {
         requireNonNull(path, "path");
-        return findNode(root, path, 0, exact);
+        return findFirstNode(root, path, 0, exact, new IntHolder());
     }
 
     /**
      * Finds a {@link Node} which is mapped to the given {@code path}. It is recursively called by itself
      * to visit the children of the given node. Returns {@code null} if there is no {@link Node} to find.
      */
     @Nullable
-    private Node<V> findNode(Node<V> node, String path, int begin, boolean exact) {
-        final int next;
+    private Node<V> findFirstNode(Node<V> node, String path, int begin, boolean exact, IntHolder nextHolder) {
+        final Node<V> checked = checkNode(node, path, begin, exact, nextHolder);
+        if (checked != continueWalking()) {
+            return checked;
+        }
+
+        // The path is not matched to this node, but it is possible to be matched on my children
+        // because the path starts with the path of this node. So we need to visit children as the
+        // following sequences:
+        //  - The child which is able to consume the next character of the path.
+        //  - The child which has a path variable.
+        //  - The child which is able to consume every remaining path. (catch-all)
+        final int next = nextHolder.value;
+        Node<V> child = node.child(path.charAt(next));
+        if (child != null) {
+            final Node<V> found = findFirstNode(child, path, next, exact, nextHolder);
+            if (found != null) {
+                return found;
+            }
+        }
+        child = node.parameterChild();
+        if (child != null) {
+            final Node<V> found = findFirstNode(child, path, next, exact, nextHolder);
+            if (found != null) {
+                return found;
+            }
+        }
+        return node.catchAllChild();
+    }
+
+    private List<Node<V>> findAllNodes(String path, boolean exact) {
+        return findAllNodes(root, path, 0, exact, new IntHolder());
+    }
+
+    private List<Node<V>> findAllNodes(Node<V> node, String path, int begin, boolean exact,
+                                       IntHolder nextHolder) {
+        final Node<V> checked = checkNode(node, path, begin, exact, nextHolder);
+        if (checked != continueWalking()) {
+            return checked == null ? ImmutableList.of() : ImmutableList.of(checked);
+        }
+
+        final int next = nextHolder.value;
+        final ImmutableList.Builder<Node<V>> builder = ImmutableList.builder();
```

Comment:
Since this is a `private` method, might be better to accept `ImmutableList.Builder` as argument instead of return, so we don't have to create a new one at every recursive invocation of `findAllNodes`

Reply:
Fixed :-)